crates/experimentation_client/src/lib.rs

```diff
@@ -212,36 +250,67 @@ impl Client {
             })
             .collect()
     }
+}
 
-    // decide which variant to return among all applicable experiments
-    fn decide_variant(
-        &self,
-        traffic: u8,
-        applicable_variants: Variants,
-        toss: i8,
-    ) -> Result<Option<Variant>, String> {
-        if toss < 0 {
-            for variant in applicable_variants.iter() {
-                if variant.variant_type == VariantType::EXPERIMENTAL {
-                    return Ok(Some(variant.clone()));
-                }
-            }
-        }
-        let variant_count = applicable_variants.len() as u8;
-        let range = (traffic * variant_count) as i32;
-        if (toss as i32) >= range {
-            return Ok(None);
-        }
-        let buckets = (1..=variant_count)
-            .map(|i| (traffic * i) as i8)
-            .collect::<Vec<i8>>();
-        let index = buckets
-            .into_iter()
-            .position(|x| toss < x)
-            .ok_or_else(|| "Unable to fetch variant's index".to_string())
-            .map_err_to_string()?;
-        Ok(applicable_variants.get(index).cloned())
-    }
+pub fn get_applicable_buckets_from_group(
+    experiment_groups: &[ExperimentGroup],
+    context: &Value,
+    identifier: &str,
+) -> Vec<(usize, Bucket)> {
+    experiment_groups
+        .iter()
+        .filter_map(|exp_group| {
+            let hashed_percentage = calculate_bucket_index(identifier, &exp_group.id);
+            let exp_context: Map<String, Value> = exp_group.context.clone().into();
+            let res = (exp_context.is_empty()
+                || jsonlogic::apply(&Value::Object(exp_context), context)
+                    == Ok(Value::Bool(true)))
+                && *exp_group.traffic_percentage >= hashed_percentage as u8;
+            let exp_toss =
+                (hashed_percentage * 100) / *exp_group.traffic_percentage as usize;
+
+            res.then_some(
+                exp_group
+                    .buckets
+                    .get(hashed_percentage)
+                    .and_then(Clone::clone),
+            )
+            .flatten()
+            .map(|b| (exp_toss, b))
+        })
+        .collect()
+}
+
+pub fn get_applicable_variants_from_group_response(
+    experiments: &[ExperimentResponse],
+    context: &Value,
+    bucket_response: &[(usize, Bucket)],
+) -> Vec<String> {
+    experiments
```

Comment:
can we make it efficient ? currently its num_of_experiment * satisfied_buckets 
can we first create a hashmap of experiment and then iter on satisfied_buckets and validate the context